src/db.rs

```diff
@@ -1,67 +1,111 @@
-use crate::dir::Dir;
-use crate::types::{Epoch, Rank};
+use crate::config;
+use crate::dir::{Dir, Epoch, Rank};
+
 use anyhow::{anyhow, bail, Context, Result};
-use std::fs::{self, File};
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+use std::cmp::Ordering;
+use std::fs::{self, File, OpenOptions};
 use std::io::{self, BufRead, BufReader, BufWriter};
 use std::path::{Path, PathBuf};
 
+pub use i32 as DBVersion;
+
 pub struct DB {
-    path: PathBuf,
-    dirs: Vec<Dir>,
+    data: DBData,
     modified: bool,
+    path: PathBuf,
+    path_old: Option<PathBuf>,
 }
 
 impl DB {
     pub fn open<P: AsRef<Path>>(path: P) -> Result<DB> {
-        let path = path.as_ref().to_path_buf();
-
-        let dirs = match File::open(&path) {
+        let data = match File::open(&path) {
             Ok(file) => {
                 let reader = BufReader::new(&file);
-                bincode::deserialize_from(reader)
-                    .with_context(|| anyhow!("could not deserialize database"))?
+                bincode::config()
+                    .limit(config::DB_MAX_SIZE)
+                    .deserialize_from(reader)
+                    .context("could not deserialize database")?
             }
             Err(err) => match err.kind() {
-                io::ErrorKind::NotFound => Vec::<Dir>::new(),
-                _ => return Err(err).with_context(|| anyhow!("could not open database file")),
+                io::ErrorKind::NotFound => DBData::default(),
+                _ => return Err(err).context("could not open database file"),
             },
         };
 
+        if data.version != config::DB_VERSION {
+            bail!("this database version ({}) is unsupported", data.version);
```

Comment:
Get right to the point.

```suggestion
            bail!("database version '{}' is unsupported", data.version);
```